cpp/unreal_plugins/UrdfRobot/Source/UrdfRobot/UrdfRobotComponent.cpp

```diff
@@ -0,0 +1,335 @@
+    // Need to explicitly set this up so that the component hierarchy is well-defined.
+    input_action_component_->SetupAttachment(this);
+}
+
+UUrdfRobotComponent::~UUrdfRobotComponent()
+{
+    SP_LOG_CURRENT_FUNCTION();
+
+    // Objects created with CreateDefaultSubobject, DuplicateObject, LoadObject, NewObject don't need to be cleaned up explicitly.
+
+    LinkComponents.Empty();
+    JointComponents.Empty();
+
+    link_components_.clear();
+    joint_components_.clear();
+
+    SP_ASSERT(input_action_component_);
+    input_action_component_ = nullptr;
+}
+
+void UUrdfRobotComponent::BeginPlay()
+{
+    USceneComponent::BeginPlay();
+
+    // Get player input actions from the config system if it is initialized, otherwise use hard-coded keyboard actions, which
+    // can be useful for debugging.
+    std::map<std::string, std::map<std::string, std::vector<double>>> input_actions;
+    if (Config::s_initialized_) {
+        input_actions = Config::get<std::map<std::string, std::map<std::string, std::vector<double>>>>("URDF_ROBOT.URDF_ROBOT_COMPONENT.INPUT_ACTIONS");
+    } else {
+        input_actions = DEFAULT_INPUT_ACTIONS;
+    }
+
+    input_action_component_->bindInputActions(input_actions);
+    input_action_component_->apply_input_action_func_ = [this, input_actions](const std::string& key) -> void {
+        if (EnableKeyboardControl) {
+            // only assert if we're not in the editor
+            bool assert_if_action_is_inconsistent_with_joint = !WITH_EDITOR;
+            bool assert_if_joint_not_found = !WITH_EDITOR;
+            applyAction(input_actions.at(key), assert_if_joint_not_found, assert_if_action_is_inconsistent_with_joint);
+        }
+    };
+}
+
+void UUrdfRobotComponent::TickComponent(float DeltaTime, enum ELevelTick TickType, FActorComponentTickFunction* ThisTickFunction)
+{
+    USceneComponent::TickComponent(DeltaTime, TickType, ThisTickFunction);
+
+    if (!ticked_once_) {
+        SP_LOG_CURRENT_FUNCTION();
+        initializeDeferred();
+        ticked_once_ = true;
+    }
+
+    // Update this component's pose to match root link component
+    SP_ASSERT(root_link_component_);
+    bool sweep = false;
+    FHitResult* hit_result = nullptr;
+    SetWorldLocationAndRotation(root_link_component_->GetComponentLocation(), root_link_component_->GetComponentRotation(), sweep, hit_result, ETeleportType::None);
+}
+
+std::map<std::string, ArrayDesc> UUrdfRobotComponent::getActionSpace() const
+{
+    std::map<std::string, ArrayDesc> action_space;
+
+    if (Std::contains(action_components_, "control_joints")) {
+        for (auto joint_component : joint_components_) {
+            action_space.merge(joint_component.second->getActionSpace());
+        }
+    }
+
+    // remove empty map entry
+    action_space.erase("");
+
+    return action_space;
+}
+
+std::map<std::string, ArrayDesc> UUrdfRobotComponent::getObservationSpace() const
+{
+    std::map<std::string, ArrayDesc> observation_space;
```

Comment:
- Delegate this entire interface to the individual links and joints.
- 
```
    std::map<std::string, ArrayDesc> observation_space;

    if (Std::contains(observation_components_, "link_configurations")) {
        for (auto link_component : link_components_) {
            observation_space.merge(link_component->getObservationSpace());
        }
    }

    if (Std::contains(observation_components_, "joint_configurations")) {
        for (auto joint_component : joint_components_) {
            observation_space.merge(joint_component->getObservationSpace());
        }
    }

    return observation_space;
```

Reply:
Updated code to delegate these function calls except `applyAction()`.

Reply:
I agree that `applyAction` is a somewhat special case because it requires iterating over the action and finding the appropriate object to delegate to. I added recommendations for how to implement below.